ISPN-11744 Fix ManagedConnectionOperations

tristantarrant opened the pull request:
https://issues.redhat.com/browse/ISPN-11744

diegolovison commented:
Looks good to me! :)

alanfx commented:
Hey @tristantarrant will this still allow using additional databases from DBAllocator?

tristantarrant commented:
Yes, you will however need to pass -Dorg.infinispan.server.persistence.PersistenceIT=/path/to/an/xml/with/the/correct/datasources.xml

alanfx commented on server/testdriver/core/src/main/java/org/infinispan/server/test/core/InfinispanServerTestConfiguration.java:
```diff
@@ -18,10 +20,11 @@
    public InfinispanServerTestConfiguration(String configurationFile, int numServers, ServerRunMode runMode,
                                             Properties properties, String[] mavenArtifacts, JavaArchive[] archives,
-                                            boolean jmx, boolean parallelStartup) {
+                                            boolean jmx, boolean parallelStartup, List<InfinispanServerListener> listeners) {
```
Nit: add this parameter on a new line

alanfx commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,73 @@
+      } else {
+         this.databaseTypes = databaseTypes;
+      }
+
```
Nit: Extra blank line

alanfx commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,73 @@
+         database.start();
+         log.infof("Started database: %s", database.getType());
+         databases.put(databaseTypes[i], database);
+        driver.getConfiguration().properties().put("org.infinispan.server.test."+ database.getType()+".jdbcUrl", database.jdbcUrl());
```
The formatting looks weird in this block?

alanfx commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,73 @@
+      log.info("Stopping databases");
+      databases.values().forEach(Database::stop);
+      log.info("Stopped databases");
+
```
Nit: Extra blank line

gustavolira commented:
> org.infinispan.server.persistence

README.md is using an outdated paremter, can you update README.md with this new parameter?

gustavolira commented:
@tristantarrant external databases for `PooledConnectionOperations` is not working, we should be able to pass the parameters like jdbcurl, user, password so the code could use these parameters instead of get from properties file. Currently even if we pass `-Dorg.infinispan.test.server.jdbc.databases=mysql -Ddatabase.mode=EXTERNAL` the code will get `mysql.properties` which have `database.mode=CONTAINER` so container will be used and a new instance of `ContainerDatabase` will be created.
If you want we can integrate this PR and them I can submit another PR focusing on external databases

gustavolira commented:
> Yes, you will however need to pass `-Dorg.infinispan.server.persistence.PersistenceIT=/path/to/an/xml/with/the/correct/datasources.xml`

Using `mvn verify -pl server/tests -Dit.test=ManagedConnectionOperations -Dorg.infinispan.server.persistence.PersistenceIT=/home/gustavo/database.xml` I'm receiving 
```
java.nio.file.NoSuchFileException: /home/gustavo/snap/gnome-calculator/common/.cache/gio-modules/libdconfsettings.so
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:548)
	at java.base/sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258)
	at java.base/java.nio.file.Files.copy(Files.java:1294)
	at org.infinispan.commons.util.Util$1.visitFile(Util.java:1158)
	at org.infinispan.commons.util.Util$1.visitFile(Util.java:1139)
	at java.base/java.nio.file.Files.walkFileTree(Files.java:2724)
	at org.infinispan.commons.util.Util.recursiveDirectoryCopy(Util.java:1139)
	at org.infinispan.server.test.core.AbstractInfinispanServerDriver.prepare(AbstractInfinispanServerDriver.java:99)
	... 23 more

```

tristantarrant commented:
I have fixed the nitpicks and cleaned up the properties. You can now pass a `org.infinispan.test.server.database.properties.path` which should point to a directory of _type_.properties. I have no idea why you are getting that @gustavolira

gustavolira commented:
LGTM

gustavolira commented on server/testdriver/README.md:
```diff
@@ -113,13 +113,12 @@ The following is a list of properties which affect the build:
-
+* `org.infinispan.test.server.database.types` comma-separated list of database types to be used during persistence tests.
+* `org.infinispan.test.server.database.properties.path` a path to a directory containing property files with connection details for each database to be tested.
+* `org.infinispan.server.test.database.type.username` username to use for a specific database type 
```
Looks like `org.infinispan.server.test.database.type.*` is no longer being used since you can pass these properties through `org.infinispan.test.server.database.properties.path`
LGTM despite that

tristantarrant replied:
I'm using them in the `type.properties` files. They are not essential, but demonstrate how to add extra configurability

pruivo replied:
ah... I totally miss that. can you put the word `type` between `<>`  to make it clear?

pruivo commented on server/runtime/src/main/java/org/infinispan/server/configuration/DataSourcesConfigurationBuilder.java:
```diff
@@ -1,28 +1,39 @@
 
    DataSourceConfigurationBuilder dataSource(String name, String jndiName) {
+      if (dataSources.containsKey(name)) {
+         throw Server.log.duplicateDataSource(name);
```
Similar to JDBC [1] you could log the exception as `CONFIG` logger.

[1] https://github.com/infinispan/infinispan/blob/master/persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/logging/Log.java#L30

pruivo replied:
or not... it seems to be using the `SERVER` category for this in other places...

tristantarrant replied:
This is a server configuration exception. It's fine as it is

pruivo commented on server/runtime/src/main/java/org/infinispan/server/configuration/DataSourcesConfigurationBuilder.java:
```diff
@@ -1,28 +1,39 @@
    private final ServerConfigurationBuilder server;
 
    private Map<String, DataSourceConfigurationBuilder> dataSources = new LinkedHashMap<>(2);
+   private Set<String> jndiNames = new HashSet<>(2);
```
nitpick: both `jndiNames` and  `dataSources` can be final.

pruivo commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,74 @@
+   @Override
+   public void before(InfinispanServerDriver driver) {
+      databases = new LinkedHashMap<>(databaseTypes.length);
+      for (int i = 0; i < databaseTypes.length; i++) {
```
```suggestion
      for (String databaseType : databaseTypes) {
```

pruivo commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,74 @@
+         log.infof("Starting database: %s", database.getType());
+         database.start();
+         log.infof("Started database: %s", database.getType());
+         databases.put(databaseTypes[i], database);
```
shouldn't check for duplicates before starting?

pruivo commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/DatabaseServerListener.java:
```diff
@@ -0,0 +1,74 @@
+
+   private Database initDatabase(String databaseType) {
+      String property = System.getProperty(DATABASE_PROPERTIES_PATH);
+      try (InputStream inputStream = property != null ? Files.newInputStream(Paths.get(property).resolve(databaseType + ".properties")) : getClass().getResourceAsStream(String.format("/database/%s.properties", databaseType))) {
```
will `/database/%s.properties`  work on windows?

tristantarrant replied:
Yes, because it's reading it from the classpath here.

pruivo commented on server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRuleBuilder.java:
```diff
@@ -22,6 +25,7 @@
    private JavaArchive[] archives;
    private boolean jmx;
    private boolean parallelStartup = true;
+   private List<InfinispanServerListener> listeners = new ArrayList<>();
```
nitpick: `final`

pruivo commented on server/tests/src/test/resources/configuration/datasources/default.xml:
```diff
@@ -0,0 +1,29 @@
+      <connection-pool max-size="10" background-validation="1000" idle-removal="1" initial-size="1" leak-detection="10000"/>
+   </data-source>
+   <data-source name="mysql" jndi-name="jdbc/mysql" statistics="true">
+      <!-- The URL will be filled in by -->
```
remove comment?

pruivo commented:
> > Yes, you will however need to pass `-Dorg.infinispan.server.persistence.PersistenceIT=/path/to/an/xml/with/the/correct/datasources.xml`
> 
> Using `mvn verify -pl server/tests -Dit.test=ManagedConnectionOperations -Dorg.infinispan.server.persistence.PersistenceIT=/home/gustavo/database.xml` I'm receiving
> 
> ```
> java.nio.file.NoSuchFileException: /home/gustavo/snap/gnome-calculator/common/.cache/gio-modules/libdconfsettings.so
> 	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
> 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
> 	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
> 	at java.base/sun.nio.fs.UnixCopyFile.copy(UnixCopyFile.java:548)
> 	at java.base/sun.nio.fs.UnixFileSystemProvider.copy(UnixFileSystemProvider.java:258)
> 	at java.base/java.nio.file.Files.copy(Files.java:1294)
> 	at org.infinispan.commons.util.Util$1.visitFile(Util.java:1158)
> 	at org.infinispan.commons.util.Util$1.visitFile(Util.java:1139)
> 	at java.base/java.nio.file.Files.walkFileTree(Files.java:2724)
> 	at org.infinispan.commons.util.Util.recursiveDirectoryCopy(Util.java:1139)
> 	at org.infinispan.server.test.core.AbstractInfinispanServerDriver.prepare(AbstractInfinispanServerDriver.java:99)
> 	... 23 more
> ```

@gustavolira the driver copies everything from the folder where the config is stored. That's why it is copying your `/home/gustavo/*`. You need to be careful with your paths :) 
```
// Recursively copy the contents of the directory containing the configuration file to the test target
Util.recursiveDirectoryCopy(configurationFilePath.getParent(), confDir.toPath());
```

pruivo commented:
integrated! thanks @tristantarrant !
